### PR TITLE
Base Class GG-Lensing refactor

### PR DESCRIPTION
**Pull request by padma18-vb**
In this branch, I'll be working on building in base classes for objects involved in creating lensed populations and refactoring the current gg-lensing branch in sim-pipeline. All commits I make should keep showing up, so I'll check in if I have any questions about some section of the code and treat it as a venue to keep everyone updated on my progress!

Below is an ongoing list of changes I have made so far:

- Removed galsim usage due to installation issues (FFT wrapper tool configuration failed)
- Added deflector_base.py that will store properties of deflectors - implemented it as an Abstract Base Class (ABC)
- gg_lens and OM10AGN_lens should now inherit from deflector_base.py
- added base classes for lensed systems, lens samples and deflectors

@sibirrer @nkhadka21 Please let me know if you have any suggestions for what I can change!

### CODE REVIEW

File: test.txt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+
```
Comment (sibirrer) [inline test.txt:1]: is this file required?

File: sim_pipeline/image_simulation.py [hunk 2]
```diff
@@ -71,17 +71,17 @@ def gsobj(image, pix_scale, flux):
...
+    # gso = galsim.InterpolatedImage(galsim.Image(image), scale = pix_scale, flux = flux)
+    # return gso
 
 def gsobj_true_flux(image, pix_scale):
     """
     :param image: image that need to be interpolated
     :param pix_scale: pixel scale to be asigned to the interpolated image
     :param flux: flux value to be asigned to the interpolated image
     :returns : interpolated image with specified pixel scale and flux"""
-    gso_test = galsim.InterpolatedImage(galsim.Image(image), scale = pix_scale, normalization='flux')
```
Comment (sibirrer) [inline sim_pipeline/image_simulation.py:83]: this function is required for the RSP source injection. So galsim is somehow required

File: sim_pipeline/image_simulation.py [hunk 3]
```diff
@@ -71,17 +71,17 @@ def gsobj(image, pix_scale, flux):
     :param pix_scale: pixel scale to be asigned to the interpolated image
     :param flux: flux value to be asigned to the interpolated image
     :returns : interpolated image with specified pixel scale and flux"""
-    gso = galsim.InterpolatedImage(galsim.Image(image), scale = pix_scale, flux = flux)
```
Comment (sibirrer) [inline sim_pipeline/image_simulation.py:74]: I would not change these parts of the code since @nkhadka21 is working on and with them

File: sim_pipeline/Deflector/deflector_base.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+        """
+
+        :param band: imaging band, if =None, will result in un-normalized amplitudes
+        :type band: string or None
+        :return: lenstronomy model and parameter conventions
+
+        """
+        pass
+    
+
```
Comment (sibirrer) [inline sim_pipeline/Deflector/deflector_base.py:195]: can you remove the black lines (except one blank line after the last routine

File: sim_pipeline/Deflector/deflector_base.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+        """
+        pass
+
+
+    @abstractmethod
+    def deflector_light_model_lenstronomy(self):
+        """
+        returns lens model instance and parameters in lenstronomy conventions
+
+        :return: lens_model_list, kwargs_lens
```
Comment (sibirrer) [inline sim_pipeline/Deflector/deflector_base.py:178]: I guess this is light_model_list

File: sim_pipeline/Deflector/deflector_base.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+        :return: stellar mass of deflector
+        """
+        pass
+
+    def deflector_velocity_dispersion(self):
+        """
+
+        :return: velocity dispersion [km/s]
+        """
+        return self._lens_dict['vel_disp']
```
Comment (sibirrer) [inline sim_pipeline/Deflector/deflector_base.py:94]: does this object exist? Also, can you add tests for the base class?

File: sim_pipeline/Deflector/deflector_base.py [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+        :return: [x_pos, y_pos]
+        """
+
+    @abstractmethod
+    def image_positions(self):
+        """
+        returns image positions by solving the lens equation, these are either the centers of the extended source, or
+        the point sources in case of (added) point-like sources, such as quasars or SNe.
+
+        :return: x-pos, y-pos
```
Comment (sibirrer) [inline sim_pipeline/Deflector/deflector_base.py:44]: do we want to specify that the positions are in arrays (as they can be multiple images)

File: sim_pipeline/Deflector/OM10AGNlens.py [hunk 8]
```diff
@@ -0,0 +1,24 @@
+from gg_lens import GGLens
+from sim_pipeline.Deflector.deflector_base import LensBase
```
Comment (sibirrer) [inline sim_pipeline/Deflector/OM10AGNlens.py:2]: is this DeflectorBase or LensBase? I don't see LensBase in the pull request

File: sim_pipeline/Deflector/OM10AGNlens.py [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    """
+    Class to manage individual galaxy-AGN lenses from OM10 Catalog
+    OM10 Catalog: https://github.com/drphilmarshall/OM10/tree/master
+    """
+
+    def __init__(self, source_dict, deflector_dict, cosmo,
+                 source_type='point_source',
+                 test_area=4 * np.pi):
+        
+        super(GGLens, self).__init__(source_dict, deflector_dict, cosmo,
```
Comment (sibirrer) [inline sim_pipeline/Deflector/OM10AGNlens.py:15]: Does this work? GGLens is not defined

**Review by sibirrer — CHANGES_REQUESTED** [review]
Thanks for the PR @padma18-vb !
I made some comments and ask you to write some unit tests for the classes you introduced.

File: test.txt (hunk 1 above)
Comment (padma18-vb) [reply test.txt:1]: No, just deleted it, thanks!

File: sim_pipeline/image_simulation.py (hunk 3 above)
Comment (padma18-vb) [reply sim_pipeline/image_simulation.py:74]: Got it! Will find a work around to the galsim installation issues.

File: sim_pipeline/Deflector/deflector_base.py (hunk 5 above)
Comment (padma18-vb) [reply sim_pipeline/Deflector/deflector_base.py:178]: Yes, thanks for catching that!

File: sim_pipeline/Deflector/deflector_base.py (hunk 7 above)
Comment (padma18-vb) [reply sim_pipeline/Deflector/deflector_base.py:44]: Good idea - updated!

File: sim_pipeline/Deflector/OM10AGNlens.py (hunk 8 above)
Comment (padma18-vb) [reply sim_pipeline/Deflector/OM10AGNlens.py:2]: DeflectorBase, that was a residual from a name change. Thanks!

File: sim_pipeline/Deflector/OM10AGNlens.py (hunk 9 above)
Comment (padma18-vb) [reply sim_pipeline/Deflector/OM10AGNlens.py:15]: This also supposed to inherit from DeflectorBase. Updated!

**Comment by sibirrer** [comment]
Hi @padma18-vb, just checking in on whether you have questions or road blocks with this PR :)

**Comment by padma18-vb** [comment]
Hi Simon!

I've been at the SLAC Summer Institute so I didn't get the time to work on
this over the past 5 days - no technical hold-ups, I'll commit my work as I
make more progress!

**Review by sibirrer — CHANGES_REQUESTED** [review]
Thank you @padma18-vb !
I see quite some .idea files in the commit which you should take out (might be cached files etc). Can you also re-direct your PR to the main branch? Thank you very much!

File: sim_pipeline/Sources/source_base.py [hunk 10]
```diff
@@ -1,7 +1,16 @@
+<<<<<<< HEAD
+from abc import ABC, abstractmethod
+
+class SourceBase(ABC):
+    """
+    Base class with functions all source classes must have to be able to render populations
+    """
+=======
```
Comment (sibirrer) [inline sim_pipeline/Sources/source_base.py:8]: it looks like you have accepted both versions and also the patterns of the merge conflict '======' and '>>>>>>/

**Review by sibirrer — CHANGES_REQUESTED** [review]
Hi @padma18-vb, thanks a lot! It looks like you have accepted either both merge conflicts AND the distinguishing lines '====' and. '>>>>'. If you remove the duplications and these lines, I hope the PR tests will be running

File: sim_pipeline/Sources/source_base.py (hunk 10 above)
Comment (padma18-vb) [reply sim_pipeline/Sources/source_base.py:8]: Ah, got it! I missed a couple of those fixes, will re-commit!

**Comment by sibirrer** [comment]
@padma18-vb , looks great. There are a few other places the re-naming has to propagate. You can look at the tests of the CI (continuous Integration) and the error messages there. We are getting there.

File: sim_pipeline/galaxy_galaxy_lens.py [hunk 11]
```diff
@@ -415,13 +419,97 @@ def lenstronomy_kwargs(self, band=None):
...
+
+        if band is None:
+            mag_lens = 1
+        else:
+            mag_lens = self.deflector_magnitude(band)
+        kwargs_lens_light = [
+            {
+                "magnitude": mag_lens,
+                "R_sersic": size_lens_arcsec,
+                "n_sersic": float(self._source_dict["n_sersic"]),
```
Comment (sibirrer) [inline sim_pipeline/galaxy_galaxy_lens.py:495]: I don't know why this is new code, but we fixed a bug here switching _source_dict to _deflector_dict

File: sim_pipeline/galaxy_galaxy_lens_pop.py [hunk 12]
```diff
@@ -132,29 +127,29 @@ def select_lens_at_random(self, **kwargs_lens_cut):
...
 
-    def deflector_number(self):
+    def potential_deflector_number(self):
         """Number of potential deflectors (meaning all objects with mass that are being
         considered to have potential sources behind them)
 
         :return: number of potential deflectors
         """
-        return self._lens_galaxies.deflector_number()
+        return len(self._lens_galaxies)
```
Comment (sibirrer) [inline sim_pipeline/galaxy_galaxy_lens_pop.py:136]: is this function each time going to re-compute a quantity? There are certain functions that need repeated access to this number such that if this number is saved, can avoid unnecessary calculations

File: sim_pipeline/galaxy_galaxy_lens_pop.py [hunk 13]
```diff
@@ -132,29 +127,29 @@ def select_lens_at_random(self, **kwargs_lens_cut):
...
-    def source_number(self):
+    def potential_source_number(self):
         """Number of sources that are being considered to be placed in the sky area
         potentially aligned behind deflectors.
 
-        :return: number of sources
+        :return: number of potential sources
         """
-        return self._sources.source_number()
+        return len(self._sources)
```
Comment (sibirrer) [inline sim_pipeline/galaxy_galaxy_lens_pop.py:144]: same as from the deflector numbers. It can be ok to duplicate access points in different classes as long as the underlying function is the same to be called

File: test.txt (hunk 1 above)
Comment (sibirrer) [inline test.txt:1]: I guess this file can be deleted?

**Review by sibirrer — CHANGES_REQUESTED** [review]
Looks very good. I have some minor questions and a request for a bug fix that was introduced somewhere else

File: sim_pipeline/galaxy_galaxy_lens.py (hunk 11 above)
Comment (padma18-vb) [reply sim_pipeline/galaxy_galaxy_lens.py:495]: Will address these comments now and push again!

File: sim_pipeline/galaxy_galaxy_lens_pop.py (hunk 12 above)
Comment (padma18-vb) [reply sim_pipeline/galaxy_galaxy_lens_pop.py:136]: I think here, len(self._lens_galaxies) will function the same way as self._lens_galaxies.deflector_number() as both methods are just calling a precomputed number that has been stored in the _lens_galaxies object!

**Review by nkhadka21 — CHANGES_REQUESTED** [review]

File: sim_pipeline/galaxy_galaxy_lens_pop.py (hunk 13 above)
Comment (padma18-vb) [reply sim_pipeline/galaxy_galaxy_lens_pop.py:144]: Okay, this makes sense! In galaxy_galaxy_lens_pop, I renamed deflector_number() to potential_deflector_number() and source_number() to potential_source_number(). Do you recommend that I keep this name and use the same method name (i.e., potential_deflector_number and potential_source_number) for Deflector and Source objects instead of len(...)?

Comment (sibirrer) [reply sim_pipeline/galaxy_galaxy_lens_pop.py:144]: hmm, I am a bit torn here. 'potential' is also used for lensing potential in certain terms. it might be easier to keep them without 'potential', or 'all' or 'unlensed'. I leave it up to @nkhadka21 and you (@padma18-vb ) :)

File: sim_pipeline/galaxy_galaxy_lens_pop.py (hunk 12 above)
Comment (sibirrer) [reply sim_pipeline/galaxy_galaxy_lens_pop.py:136]: sounds good! not so used to overwrite class methods but good coding

**Comment by sibirrer** [comment]
@padma18-vb: there are some abstract classes that are not fully tested. you can have a look at the coverage report above. Maybe there is a chance for better testing (there is never enough testing...)

File: sim_pipeline/galaxy_galaxy_lens_pop.py (hunk 13 above)
Comment (nkhadka21) [reply sim_pipeline/galaxy_galaxy_lens_pop.py:144]: I prefer to keep them without potential. It is also easier to keep track of test functions with fewer changes.

File: sim_pipeline/galaxy_galaxy_lens_pop.py (hunk 12 above)
Comment (sibirrer) [reply sim_pipeline/galaxy_galaxy_lens_pop.py:136]: @padma18-vb  I reconsidered this point. I am not such a fan of overwriting __len__ functions. I am also not sure whether len() would work anymore if not defined for a class. Also, for backwards compatibility, having the function as it was does not hurt. It's mainly about backwards compatibility and it seems this change would not really be necessary to add your functionality

File: sim_pipeline/galaxy_galaxy_lens_pop.py [hunk 14]
```diff
@@ -110,11 +105,11 @@ def __init__(
         self.cosmo = cosmo
         self.f_sky = sky_area
 
-    def select_lens_at_random(self, **kwargs_lens_cut):
+    def generate_random_lensed_system(self, **kwargs_lens_cut):
```
Comment (sibirrer) [inline sim_pipeline/galaxy_galaxy_lens_pop.py:108]: was there a specific reason to change this? The reason I am asking is as multiple other people are using the sim_pipeline this is again not backwards compatible and might create some hassle

File: data/Filters/Roman/Roman-F062.ecsv [hunk 15]
```diff
@@ -0,0 +1,85 @@
+# %ECSV 1.0
```
Comment (sibirrer) [inline data/Filters/Roman/Roman-F062.ecsv:1]: can you not add these files, as they will be generated when running the code and save disk, git and download space. I'd recommend only adding explicitly files to git that you write and want to have added

File: data/Filters/Roman/Roman-F087.ecsv [hunk 16]
```diff
@@ -0,0 +1,77 @@
+# %ECSV 1.0
```
Comment (sibirrer) [inline data/Filters/Roman/Roman-F087.ecsv:1]: remove from git please

File: data/Filters/Roman/Roman-F106.ecsv [hunk 17]
```diff
@@ -0,0 +1,90 @@
+# %ECSV 1.0
```
Comment (sibirrer) [inline data/Filters/Roman/Roman-F106.ecsv:1]: remove from git please

File: data/Filters/Roman/Roman-F129.ecsv [hunk 18]
```diff
@@ -0,0 +1,106 @@
+# %ECSV 1.0
```
Comment (sibirrer) [inline data/Filters/Roman/Roman-F129.ecsv:1]: remove from git please

File: data/Filters/Roman/Roman-F146.ecsv [hunk 19]
```diff
@@ -0,0 +1,260 @@
+# %ECSV 1.0
```
Comment (sibirrer) [inline data/Filters/Roman/Roman-F146.ecsv:1]: remove from git please

File: data/Filters/Roman/Roman-F158.ecsv [hunk 20]
```diff
@@ -0,0 +1,127 @@
+# %ECSV 1.0
```
Comment (sibirrer) [inline data/Filters/Roman/Roman-F158.ecsv:1]: remove from git please

File: data/Filters/Roman/Roman-F184.ecsv [hunk 21]
```diff
@@ -0,0 +1,116 @@
+# %ECSV 1.0
```
Comment (sibirrer) [inline data/Filters/Roman/Roman-F184.ecsv:1]: remove from git please

File: data/Filters/Roman/Roman-F213.ecsv [hunk 22]
```diff
@@ -0,0 +1,128 @@
+# %ECSV 1.0
```
Comment (sibirrer) [inline data/Filters/Roman/Roman-F213.ecsv:1]: remove from git please

File: data/Filters/Roman/Roman-Grism_0thOrder.ecsv [hunk 23]
```diff
@@ -0,0 +1,205 @@
+# %ECSV 1.0
```
Comment (sibirrer) [inline data/Filters/Roman/Roman-Grism_0thOrder.ecsv:1]: remove from git please

File: data/Filters/Roman/Roman-Grism_1stOrder.ecsv [hunk 24]
```diff
@@ -0,0 +1,205 @@
+# %ECSV 1.0
```
Comment (sibirrer) [inline data/Filters/Roman/Roman-Grism_1stOrder.ecsv:1]: remove from git please

File: data/Filters/Roman/Roman-SNPrism.ecsv [hunk 25]
```diff
@@ -0,0 +1,235 @@
+# %ECSV 1.0
```
Comment (sibirrer) [inline data/Filters/Roman/Roman-SNPrism.ecsv:1]: remove from git please

File: sim_pipeline/galaxy_galaxy_lens_pop.py (hunk 13 above)
Comment (padma18-vb) [reply sim_pipeline/galaxy_galaxy_lens_pop.py:144]: Yes, those are all valid points, will revert both methods to source_number()!

File: sim_pipeline/galaxy_galaxy_lens_pop.py (hunk 14 above)
Comment (padma18-vb) [reply sim_pipeline/galaxy_galaxy_lens_pop.py:108]: It was to make it clear to the user that they should be generating a `lensed_system` object. I thought that `select_lens_at_random` wasn't clear as to whether they should implement the method such that it returns a deflector or a galaxy_galaxy_lens type object (i.e a lensed system that has mapped a deflector to a source).

File: data/Filters/Roman/Roman-F062.ecsv (hunk 15 above)
Comment (sibirrer) [reply data/Filters/Roman/Roman-F062.ecsv:1]: yes, that could work. Generally only add the specific files that you want

File: sim_pipeline/galaxy_galaxy_lens_pop.py (hunk 14 above)
Comment (nkhadka21) [reply sim_pipeline/galaxy_galaxy_lens_pop.py:108]: Hi @padma18-vb, eventually we are going to use lens and LensPop as a general term for all lensed system and lens system population respectively. So, I think lens will make more sense then!

Comment (padma18-vb) [reply sim_pipeline/galaxy_galaxy_lens_pop.py:108]: Ah, okay! In that case, I think I will rename the base classes as well:

- `LensedSystem` --> `Lens`
- `LensedPopulation` --> `LensPop`

Does that sound okay?

Comment (sibirrer) [reply sim_pipeline/galaxy_galaxy_lens_pop.py:108]: Maybe we should wait re-naming them, certainly Lens and Deflector could be ambiguous for some users, which we should avoid

Comment (nkhadka21) [reply sim_pipeline/galaxy_galaxy_lens_pop.py:108]: Yes, we should wait for them.

**Review by sibirrer — APPROVED** [review]
Thank you @padma18-vb !
We will change the naming of LensedSystem() class after the fact to LensedSystemBase() as this is a skeleton class. I hope you don't mind, just be aware of it.
